### PR TITLE
Support for snippet tags other than script

### PR DESCRIPTION
**Pull request by Skejven**
## Description
Support for snippet tags other than `script`.

## Motivation and Context
There should be more freedom when it comes to defining dynamic Knot.x snippets. Current solution supports only snippets defined inside `<script>` tag. While it suits most of the cases there would be few that requires change of the snippet tag name to some custom one because of how Knot.x finds snippets (Knot.x just scans the markup for the opening of snippet tag and the first occurrence of end of the tag).
You may find more motivation in #107 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](https://github.com/Cognifide/knotx/blob/master/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: documentation/src/main/wiki/Assembler.md [hunk 1]
```diff
@@ -130,8 +133,14 @@ Detailed description of each configuration option is described in the next subse
...
-|-------:                     |:-------:                            |:-------:       |-------|
-| `address`                   | `String`                            | &#10004;       | Event bus address of the Fragment Assembler verticle. |
-| `unprocessedStrategy`       | `String`                             | &#10004;       | Strategy for unprocessed Fragments (`AS_IS`, `UNWRAP`, `IGNORE`). `UNWRAP` is default strategy if no strategy defined. |
+| Name                        | Type      | Mandatory      | Description  |
+|-------:                     |:-------:  |:-------:       |-------|
+| `address`                   | `String`  | &#10004;       | Event bus address of the Fragment Assembler verticle. |
+| `unprocessedStrategy`       | `String`  | &#10004;       | Strategy for unprocessed Fragments (`AS_IS`, `UNWRAP`, `IGNORE`). `UNWRAP` is default strategy if no strategy defined. |
+| `snippetTagName`            | `String`  | &#10004;       | The name of a tag that will be recognised as a Knot.x snippet. Remember to update [[Splitter configuration\|Splitter#how-to-configure]] |
 
+**Important - whenever you change `snippetTagName` to custom one remember that Knot.x splits
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/Assembler.md]: I'd rephrase a little bit the notice, just to warn user about what not to do in the first sentence:
```
Important - When specifying `snippetTagName` remember to not use standard HTML tags like `div`, `span`, etc.
Knot.x splits an HTML into fragments by parsing it as String to get best possible performance. 
It simply search text for the opening and first matching closing tag. It does not analyse the text as HTML. So, if you use `div` as fragmentTagName, and inside your will use multiple `div` tags too, then it will not pick the one that matches first opening, instead it will get the fragment up to the first closing `div` tag. It will result in a broken HTML structure.
```

File: documentation/src/main/wiki/Splitter.md [hunk 2]
```diff
@@ -57,48 +66,62 @@ and splits Template into three following Fragments:
...
 
 ### Splitter config
 
-| Name                        | Type                                | Mandatory      | Description  |
-|-------:                     |:-------:                            |:-------:       |-------|
-| `address`                   | `String`                            | &#10004;       | Event bus address of the Splitter verticle. |
+| Name                        | Type         | Mandatory      | Description  |
+|-------:                     |:-------:     |:-------:       |-------|
+| `address`                   | `String`     | &#10004;       | Event bus address of the Splitter verticle. |
+| `snippetTagName`            | `String`     | &#10004;       | The name of a tag that will be recognised as a Knot.x snippet. Remember to update [[Assembler configuration\|Assembler#how-to-configure]] |
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/Splitter.md]: specify what's the default value if not specified

File: documentation/src/main/wiki/Splitter.md [hunk 3]
```diff
@@ -57,48 +66,62 @@ and splits Template into three following Fragments:
...
+
+**Important - whenever you change `snippetTagName` to custom one remember that Knot.x splits
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/Splitter.md]: use the same warning as i suggested before

File: documentation/src/main/wiki/UpgradeNotes.md [hunk 4]
```diff
@@ -20,7 +20,10 @@ For more details see documentation sections in [[Server|Server#vertx-event-bus-d
 - [PR-372](https://github.com/Cognifide/knotx/pull/372) - Added cache for compiled Handlebars snippets, you may configure it in Handlebars config, see more in [[Handlebars Knot docs|HandlebarsKnot#how-to-configure]].
 - [PR-374](https://github.com/Cognifide/knotx/pull/374) - Enable keepAlive connection in http client options. This is important fix and we recommend to update your existing configuration of any http client and enable `keepAlive` option.
 - [PR-379](https://github.com/Cognifide/knotx/pull/379) - Added access logging capabilities to the Knotx HTTP Server. Establish standard configuration of logback logger. Check [[Configure Access Log|Server#configure-access-log]] to see how to configure access logs, and [[Logging|Logging]] to see how loggers can be configured to log to console, files for knot.x, access & netty logs.
-
+---
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/UpgradeNotes.md]: remove `----`

File: CHANGELOG.md [hunk 5]
```diff
@@ -21,7 +21,10 @@ List of changes that are finished but not yet released in any final version.
  - [PR-374](https://github.com/Cognifide/knotx/pull/374) - Enable keepAlive connection in http client options
  - [PR-380](https://github.com/Cognifide/knotx/pull/380) - Upgrade jsoup to 1.11.2
  - [PR-379](https://github.com/Cognifide/knotx/pull/379) - Added access logging capabilities to the Knotx HTTP Server. Establish standard configuration of logback logger.
- - [PR-383](https://github.com/Cognifide/knotx/pull/383) - Fix for #382 - Unhandled exception if query parameter consists of reserved characters.
+ - [PR-383](https://github.com/Cognifide/knotx/pull/383) - Fix for [#382](https://github.com/Cognifide/knotx/pull/382) - Unhandled exception if query parameter consists of reserved characters.
+---
```
[the remark this answers is not in the document]
Comment (Skejven) [reply CHANGELOG.md]: Just temporary - all tickets to this point are in Release Candidate that is currently under performance tests. I wanted to mark it.

File: knotx-knot/knotx-knot-fragment-splitter/src/test/java/io/knotx/splitter/impl/HtmlFragmentSplitterTest.java [hunk 6]
```diff
@@ -102,18 +84,55 @@ public void testOneFragmentMiddle() throws Exception {
...
+  public void split_whenCustomSnippetTagOneSnippet_expectThreeFragmentsAndSnippetInTheMiddle()
+      throws Exception {
+    String TEST_ONE_SNIPPET_MIDDLE_HTML = "test-custom-snippet-tag.html";
+    List<Fragment> testOneSnippetMiddle = new HtmlFragmentSplitter(CUSTOM_SCRIPT_TAG)
+        .split(FileReader.readText(TEST_ONE_SNIPPET_MIDDLE_HTML));
+    assertThat(testOneSnippetMiddle.size(), equalTo(3));
+    assertThat(testOneSnippetMiddle.get(0).isRaw(), equalTo(true));
+    assertThat(testOneSnippetMiddle.get(1).isRaw(), equalTo(false));
+    assertThat(testOneSnippetMiddle.get(1).content(), equalTo(
+        "<knotx:snippet data-knotx-knots=\"templating-X\" data-knotx-service=\"first-service\" type=\"text/knotx-snippet\"><h2>{{message}}</h2></knotx:snippet>"));
```
Comment (pun-ky) [inline knotx-knot/knotx-knot-fragment-splitter/src/test/java/io/knotx/splitter/impl/HtmlFragmentSplitterTest.java]: if you are using scoped `<knot:x ... >` tag name ... why attributes have prefix `data-knotx`? :) I think in that case simplification will be nice 

`<knotx:snippet knots="" service="" ...>`

Comment (Skejven) [reply knotx-knot/knotx-knot-fragment-splitter/src/test/java/io/knotx/splitter/impl/HtmlFragmentSplitterTest.java]: Currently `<script`> is still the default tag, and `data-*` naming convention fits HTML5 naming rules.

Comment (pun-ky) [reply knotx-knot/knotx-knot-fragment-splitter/src/test/java/io/knotx/splitter/impl/HtmlFragmentSplitterTest.java]: however `knotx:snippet` is not valid HTML5 tag ;)

Comment (pun-ky) [reply knotx-knot/knotx-knot-fragment-splitter/src/test/java/io/knotx/splitter/impl/HtmlFragmentSplitterTest.java]: I mean.. that short named parameters to be available only in that new `knotx:snippet` so that you could keep it backward compatible
